Maintenance: Use class extensions where appropriate

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
Closes #1350.

This PR reworks the `Utilities` class and introduces several extensions for `UILabel`, `UIView`, `UIImageView`, `UIColor`, `UIImage` and `NSString`. Some existing extensions were moved there as well, and the dedicated files for these extension got removed (image resizing and SHA256 methods).

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- You can keep it empty if it's identical to the PR title though -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Maintenance: Use class extensions where appropriate

wutschel committed:
Introduce UIImageView extensions

wutschel committed:
Introduce UIView extensions

wutschel committed:
Better parameter naming

The parameters reflect the image color for light mode and dark mode.

wutschel committed:
Refactor watched flag visibility

Convert playcount into BOOL and use this to set visibility.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1456,3 +1452,13 @@ + (void)stopPollingActiveServer {
+@implementation UILabel (Extensions)
+
+- (CGSize)getSize {
+    return [self sizeThatFits:CGSizeMake(self.frame.size.width, CGFLOAT_MAX)];
```
1. bad method name
2. as it uses nothing label-specific, it should be UIView extension instead

wutschel replied:
1. Changed to getFittingSize, hope that's ok
2. True, changed this.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1462,3 +1462,45 @@ - (CGSize)getSize {
+    self.frame = frame;
+}
+
+- (void)offsetY:(CGFloat)offset {
```
I'd call it `offsetByY`

wutschel replied:
I was having in mind "offset y by value" and then just dropped the "by" -> offsetY:value. But I can change this, if you want.

kambala-decapitator replied:
Then you can call it offsetYBy

wutschel replied:
Don't really like it, but done.

kambala-decapitator commented on XBMC Remote/Utilities.h:
```diff
@@ -145,5 +144,6 @@ typedef NS_ENUM(NSInteger, LogoBackgroundType) {
 - (void)setHeight:(CGFloat)height;
 - (void)setWidth:(CGFloat)width;
 - (void)offsetY:(CGFloat)offset;
+- (void)setX:(CGFloat)x alpha:(CGFloat)alpha;
```
since your `alpha` param is always either 0 or 1, you could instead use something like `isTransparent`

wutschel replied:
I wanted to keep `alpha` as parameter as this a generic solution, which can be used also for non 0/1 values.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1503,4 +1496,11 @@ - (void)offsetY:(CGFloat)offset {
+- (void)setX:(CGFloat)x alpha:(CGFloat)alpha {
+    CGRect frame = self.frame;
+    frame.origin.x = x;
+    self.frame = frame;
```
`[self setX:x];`

wutschel replied:
Agreed.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1500,4 +1487,14 @@ - (void)animateX:(CGFloat)x alpha:(CGFloat)alpha duration:(NSTimeInterval)second
+                     animations:^{
+        [self setOrigin:origin];
+    }
+                     completion:nil];
```
this is now basically identical to what you already have. Maybe just make a single method `- (void)defaultAnimate:(void(^)(void))animations`?

wutschel replied:
Hmm, only implementing a generic method makes this harder to read when calling. E.g.

`[activeLayoutView animateOrigin:CGPointZero duration:0.3];`

expands to sth like:

```
[activeLayoutView defaultAnimate:^{
    [activeLayoutView setOrigin:CGPointZero];
}
                         duration:0.3];
```

But I could add this as a helper and then use it inside the animation method, like

```
- (void)animateOrigin:(CGPoint)origin duration:(NSTimeInterval)seconds {
    [self defaultAnimate:^{
        [self setOrigin:origin];
    } duration:seconds];
}
```

Pushed a fixup commit for review.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1511,4 +1497,19 @@ - (UIColor*)averageColor {
 }
 
+- (UIImage*)colorizeWithColor:(UIColor*)color {
+    UIImage *image = self;
```
I'd drop this line in all methods

wutschel replied:
Yeah, this made the migration easy, but it looks weird. Changed it.

wutschel commented:
Resolved the conflict. When you're generally fine with this, I will squash most of it into 1 commit per class extension and  move the extension into dedicated .h/.m files, following the way it was recently done for the other changes.

kambala-decapitator commented on XBMC Remote/Utilities.m:
```diff
@@ -1456,3 +1193,422 @@ + (void)stopPollingActiveServer {
+}
+
+- (void)defaultAnimate:(void(^)(void))animations duration:(NSTimeInterval)seconds {
+    if (animations) {
```
Mark the param as `nullable` then

wutschel replied:
Like this?

`- (void)defaultAnimate:(nullable void(^)(void))animations duration:(NSTimeInterval)seconds {`

Just curious: I understand than `nonnull`  can help, e.g. by throwing warnings. What's the benefit of marking the callback as `nullable`?

kambala-decapitator replied:
it will also give a warning if you call it with a `nil` literal or with a `nullable` property

wutschel replied:
ok, pushed the code.

kambala-decapitator commented:
please fix build

wutschel commented:
> please fix build

Fixed, was not seeing this locally as this PR needed another rebase. Good to redo the extension now?

wutschel committed:
Move GET_ROUNDED_EDGES_RADIUS to common place

Prepares moving rounded corner implementations for UIImageView and UIImage into separate class extensions.

wutschel committed:
Introduce UIColor extensions

wutschel committed:
Introduce UIImage extensions

wutschel committed:
Introduce NSString extensions

wutschel committed:
Remove unused class extension ImageWithUIView

wutschel commented:
This was quite some nasty work. Now squashed commits into a single commit per class extension and reworked the whole implementation to use dedicated files for the class extensions. This added quite some changes to the xcode project file and many new imports to other classes.
While reviewing I came across some issues which I addressed in two fixup commits -- kept them to clearly show the difference in the implementation. In addition, I decided to remove the unused class extension `ImageWithUIView`.

wutschel commented:
Found another bug in the actors table and migrated few more frame changes to the newly introduced extensions. Reviewed this twice now after the final changes, should be good if you also agree. In this case, I would do a final squashing.

kambala-decapitator commented:
> many new imports to other classes

Like I said before, you could add extension imports to the pch

wutschel commented:
> Like I said before, you could add extension imports to the pch

Let me try that. I was only thinking of adding the extensions themselves to the pch and dropped it.

wutschel committed:
Import UIBarButtonItem+Extensions from pch

wutschel committed:
Import UILabel+Extensions from pch

wutschel committed:
Import UIViewController+Extensions from pch

wutschel committed:
Import UIImageView+WebCache.h from pch

wutschel commented:
Here we go, looks cleaner to me, also added commit to include the other extensions via the pch.

kambala-decapitator commented on XBMC Remote/Extensions/UIBarButtonItem+Extensions.h:
```diff
@@ -6,6 +6,8 @@
 //  Copyright © 2026 Team Kodi. All rights reserved.
 //
 
+@import Foundation;
```
you should be importing `UIKit` here, same in other places where you declare extensions of UIKIt classes

wutschel replied:
done

wutschel commented:
Squashed and rebased to master.
